### ISSUE TITLE
Question: Can I have multiple backward() in a training step?

### ISSUE THREAD

**Issue by 2catycm** [huggingface/accelerate#2953]
### System Info

```Shell
I have a special case of training loop, where in a single traning step I will backward multiple times.

for epoch in range(10):
      for source, targets in data:
          source = source.to(device)
          targets = targets.to(device)

          optimizer.zero_grad()

          output = model(source)
+         some submodule of model produce some local_loss according to local data and local forward result(from subsub module), which is called `local_loss.backward(retain_graph=True)`at their `forward` functions.
+         for some reason that I cannot explain, the local_loss cannot be carried out. If that is possible, I just need to let `loss=loss+local_loss`. 

          loss = F.cross_entropy(output, targets)

-         loss.backward()
+         accelerator.backward(loss)

          optimizer.step()
```

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

For example, Sparsely Gated Mixture of Experts (MoE) contains a local loss and local backward. 
Another example is Sharpness-Aware Minimization (SAM) Optimizer, which  backwards twice in the training loop. 

### Expected behavior

Since using accelerate needs to change the backward function, I wonder what is different with accelerate.backward and backward. 
To my knowledge, using raw PyTorch DDP does not need to change the backward. 
If huggingface accelerate.backward is actually designed as the "at training step end" hook,  then I don't need to use accelerate.backward in the middle steps, otherwise the logic may become wrong.
